packages/squiggle-lang/src/dist/distOperations/mixture.ts

```diff
@@ -14,43 +14,49 @@ export function mixture(
   values: [BaseDist, number][],
   { env }: { env: Env }
 ): result<BaseDist, DistError> {
-  const allValuesAreSampleSet = (v: [BaseDist, number][]) =>
-    v.every(([t]) => t instanceof SampleSetDist.SampleSetDist);
-
   if (values.length < 1) {
     return Result.Error(
       otherError("Mixture error: mixture must have at least 1 element")
     );
-  } else if (allValuesAreSampleSet(values)) {
-    const withSampleSetValues = values.map(([value, weight]) => {
-      if (value instanceof SampleSetDist.SampleSetDist) {
-        return [value, weight] as [SampleSetDist.SampleSetDist, number];
-      } else {
-        throw new Error(
-          "Mixture coding error: SampleSet expected. This should be inaccessible."
-        );
-      }
-    });
-    return SampleSetDist.mixture(withSampleSetValues, env.sampleCount);
-  } else {
-    const totalWeight = E_A_Floats.sum(values.map(([, w]) => w));
-    const properlyWeightedValues: BaseDist[] = [];
+  }
+
+  const someValuesAreSampleSets = values.some(
+    ([t]) => t instanceof SampleSetDist.SampleSetDist
+  );
+
+  if (someValuesAreSampleSets) {
+    const sampleSetValues: [SampleSetDist.SampleSetDist, number][] = [];
     for (const [dist, weight] of values) {
-      const r = scaleMultiply(dist, weight / totalWeight, { env });
-      if (!r.ok) {
-        return r;
+      if (dist instanceof SampleSetDist.SampleSetDist) {
+        sampleSetValues.push([dist, weight]);
+      } else {
+        const sampleSetResult = SampleSetDist.SampleSetDist.fromDist(dist, env);
+        if (!sampleSetResult.ok) {
+          return sampleSetResult;
+        }
+        sampleSetValues.push([sampleSetResult.value, weight]);
       }
-      properlyWeightedValues.push(r.value);
     }
+    return SampleSetDist.mixture(sampleSetValues, env.sampleCount);
+  }
 
-    let answer = properlyWeightedValues[0];
-    for (const dist of properlyWeightedValues.slice(1)) {
-      const r = binaryOperations.pointwiseAdd(answer, dist, { env });
-      if (!r.ok) {
-        return r;
-      }
-      answer = r.value;
+  const totalWeight = E_A_Floats.sum(values.map(([, w]) => w));
+  const properlyWeightedValues: BaseDist[] = [];
+  for (const [dist, weight] of values) {
+    const r = scaleMultiply(dist, weight / totalWeight, { env });
+    if (!r.ok) {
+      return r;
+    }
+    properlyWeightedValues.push(r.value);
+  }
+
+  let answer = properlyWeightedValues[0];
+  for (const dist of properlyWeightedValues.slice(1)) {
+    const r = binaryOperations.pointwiseAdd(answer, dist, { env });
```

Comment:
Minor note, but this obviously is not the ideal way to do this. We need to make a new distribution at each step. In theory, we could have a ``pointwiseSum`` fn that takes in ``n`` distributions. For later though.

Reply:
Agreed. But that's old code, I just moved it around, out of scope for this PR.

Reply:
Yea, I realize